docs/developer/architecture/user_customization_and_persistence.rst

```diff
@@ -0,0 +1,199 @@
+User Customization & Persistence
+================================
+
+Context
+-------
+
+.. todo::
+
+   This section will be handled by the requirements PR.
+
+
+Design Choices
+--------------
+
+:term:`ConfigMap` is chosen as a unified data access and storage media for user
+editable configurations in a MetalK8s cluster based on the above requirements
+for the following reasons:
+
+* Ability to support Update operations on ConfigMap's with CLI and UI easily
+  using our already existing python kubernetes module.
+* Guarantee of adaptability and ease of changing the design and implementation
+  in cases were customer needs evolve rapidly.
+* ConfigMaps stored in the :term:`etcd` database which is generally being
+  backed up. This ensures that user settings cannot be lost easily.
+
+
+How it works
+~~~~~~~~~~~~
+
+During Bootstrap, pillar values for the various services will be pre-filled
+with default values. Service pods and deployments will be configured to either
+consume configuration data directly from their ConfigMap or fallback to
+the default pillar values if the service specific ConfigMap was not found.
```

Comment:
What if we did this differently?
None of these 'user configurable' settings (at this level) should influence how the K8s cluster behaves: this should be dictated by the `BootstrapConfiguration`. As such during bootstrap, at some point *before we need any of these settings*, the K8s cluster is fully available.

Which means that at this point, we could create all these `ConfigMap`s, filled with whatever default values we choose.

At this point, we can consume them while deploying various services without any more 'defaults' defined anywhere.

Reply:
True! after a couple of discussions, an alternative we considered and updated the design docs was keeping these 'defaults' in our `default.yaml` file and then doing something like ```{%- from "metalk8s/map.jinja" import addons with context %}```. This meant in our add-ons charts we had to implement some conditional to read from the Configmaps or fall back to these defined defaults(just like pillar pre-filling idea but cleaner).

Thanks to your approach, it appears we do not even need to declare defaults in `defaults.yaml`.  Waiting to hear from others what they think.

Reply:
Like I told you Monday 3 choice:
- We have `defaults.yaml` containing this default number of replica
- we create the custos CM and "pre-fill" them **if they do not exists**
- We just put default value directly in the `sls` file

Personally suggest the second one as it will be much simpler for the user to change them if CM already exists and already contain value we want

Reply:
Option 2 looks good to me as well. A simple `kubectl delete cm ..` followed by `salt-run state.orch metalk8s.deployed` to restore defaults is acceptable I guess.